torch/_functorch/aot_autograd.py

```diff
@@ -3001,7 +3011,26 @@ def call_compiled_backward():
                         real_arg = all_args[i]
                         if not isinstance(ph_arg, torch.Tensor):
                             continue
-                        if ph_arg.stride() != real_arg.stride():
+
+                        # Comparing ph_arg.stride() with real_arg.stride() directly may
+                        # cause dynamic dimensions in ph_arg being specialized to static
+                        # value. Using the hints to avoid that.
+                        if _get_hints(ph_arg.stride()) != real_arg.stride():
+                            # Note that here we use the stride of the real tensor to
```

Comment:
CC @ezyang since I don't want to sneak in this change without your attention. I ultimately have to fallback to this 'bad' solution I initially implemented.

The other 'better' solution that decide the stride order from the real tensor and apply this order to the FakeTensor does not always work. Model like mixnet_l indeed uses some tensors with non 'dense' layout. Check the code comments here for more details. Apply a stride order to a FakeTensor will always generate a tensor with 'dense' layout.

Right now since we disable layout optimization for dynamic shape (for the other perf reason lol), it's safe to apply the stride from the real tensor directly to the fake tensor.

I'll be happy to discuss more but would like to get this fix in first to unblock.

Reply:
You can still handle the non-dense case with permutation. But the most important thing is to do the tests *in forwards*, because we have no way of adding guards if they show up in backwards right now.